### PR TITLE
[Feature]Support add/drop field for struct column(part2)

### LINKED ISSUE #46452: Support Add/drop Field for struct column
- [x] https://github.com/StarRocks/starrocks/pull/49192
- [x] https://github.com/StarRocks/starrocks/pull/48430
- [x] https://github.com/StarRocks/starrocks/pull/47590
- [x] https://github.com/StarRocks/starrocks/pull/47581
- [x] https://github.com/StarRocks/starrocks/pull/47110
- [x] https://github.com/StarRocks/starrocks/pull/47497
- [x] https://github.com/StarRocks/starrocks/pull/47206
- [x] https://github.com/StarRocks/starrocks/pull/47217
- [x] https://github.com/StarRocks/starrocks/pull/46619
- [x] https://github.com/StarRocks/starrocks/pull/47109
- [x] https://github.com/StarRocks/starrocks/pull/46451

## Feature request
Support add/drop field for struct column

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by sevev**
## Why I'm doing:
It is the second pr to support add/drop field for struct column. 

## What I'm doing:
The main changes in this PR are two fold:

1. Add the new syntax for add/drop field.
2. Support analyze add/drop field.

This pr doesn't fully implement this feature yet, and will throw not support exception when executed.

Fixes https://github.com/StarRocks/starrocks/issues/46452
## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [x] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/FieldDef.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import com.starrocks.sql.parser.NodePosition;
+
+import java.util.List;
+
+public class FieldDef implements ParseNode {
+
+    private final String fieldName;
+    private final List<String> nestedFieldNames;
+    private final TypeDef typeDef;
+    private final ColumnPosition fieldPos;
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/FieldDef.java:36]: What is the difference between fieldname and nestedfieldname. Could you add some comments for this class

Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/FieldDef.java]: > What is the difference between fieldname and nestedfieldname. Could you add some comments for this class

Done, changed in https://github.com/StarRocks/starrocks/pull/46619/commits/aade0978ca7448e2ec0ae5089e8a4ad02c0fae7e

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/FieldDef.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import com.starrocks.catalog.ScalarType;
+import com.starrocks.catalog.StructField;
+import com.starrocks.catalog.StructType;
+import com.starrocks.catalog.Type;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.sql.parser.NodePosition;
+
+import java.util.List;
+
+public class FieldDef implements ParseNode {
```
[the remark this answers is not in the document]
Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/FieldDef.java]: > Description is a syntax concept

Done, changed in https://github.com/StarRocks/starrocks/pull/46619/commits/aade0978ca7448e2ec0ae5089e8a4ad02c0fae7e

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 3]
```diff
@@ -2675,4 +2693,5 @@ nonReserved
     | WARNINGS | WEEK | WHITELIST | WORK | WRITE  | WAREHOUSE | WAREHOUSES
     | YEAR
     | DOTDOTDOT | NGRAMBF
+    | FIELD
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2696]: also add ARRAR_ELEMENT

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 4]
```diff
@@ -1006,6 +1008,22 @@ compactionClause
     : (BASE | CUMULATIVE)? COMPACT (identifier | identifierList)?
     ;
 
+subfieldName
+    : identifier | '[*]'
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1012]: replace `'[*]` with ARRAY_ELEMETN?

Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: > replace `'[*]` with ARRAY_ELEMETN?

Done, changed in https://github.com/StarRocks/starrocks/pull/46619/commits/0003760e4df6dbf046756b8683af4a9a48ff75b9

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 3 above)
Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2696]: > also add ARRAR_ELEMENT

Done, changed in https://github.com/StarRocks/starrocks/commit/0003760e4df6dbf046756b8683af4a9a48ff75b9

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableClauseVisitor.java [hunk 5]
```diff
@@ -698,6 +701,48 @@ public Void visitDropColumnClause(DropColumnClause clause, ConnectContext contex
...
     }
 
+    @Override
+    public Void visitAddFieldClause(AddFieldClause clause, ConnectContext context) {
+        String columnName = clause.getColName();
+        if (Strings.isNullOrEmpty(columnName)) {
+            throw new SemanticException(PARSER_ERROR_MSG.invalidColFormat(columnName));
+        }
+
+        if (!(table instanceof OlapTable)) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableClauseVisitor.java:711]: materialized view is not supported？

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        if (type.isArrayType()) {
+            return ((ArrayType) type).getItemType();
+        }
+
+        return null;
+    }
+
+    public void analyze(Column baseCol, boolean dropField) throws AnalysisException {
+        if (baseCol == null) {
+            throw new AnalysisException(String.format("Analyze add/drop field failed, modifyc column is not exist"));
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java:100]: modifyc -> modified?

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+                throw new AnalysisException("No filed type in field definition");
+            }
+            if (typeDef.getType().isScalarType()) {
+                final ScalarType targetType = (ScalarType) typeDef.getType();
+                if (targetType.getPrimitiveType().isStringType()) {
+                    if (targetType.getLength() <= 0) {
+                        targetType.setLength(1);
+                    }
+                }
+            }
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java:147]: child field may be struct type？

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 8]
```diff
@@ -4075,6 +4078,52 @@ public ParseNode visitDropColumnClause(StarRocksParser.DropColumnClauseContext c
...
+            StarRocksParser.IdentifierContext identifier = context.identifier(1);
+            String afterFieldName = getIdentifierName(identifier);
+            fieldPosition = new ColumnPosition(afterFieldName, createPos(identifier));
+        }
+
+        if (fieldName == null) {
+            throw new ParsingException("add field clause name is null");
+        }
+        StructFieldDesc fieldDesc = new StructFieldDesc(fieldName, parts, typeDef, fieldPosition);
+        return new AddFieldClause(columnName, fieldDesc, rollupName, getProperties(context.properties()));
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:4108]: `rollupName` is null

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableClauseVisitor.java (hunk 5 above)
Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableClauseVisitor.java:714]: > materialized view is not supported？

Not support right now. Should we support materialized view?

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java (hunk 7 above)
Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java:147]: > child field may be struct type？

Yes, child filed may be struct or array type.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java (hunk 8 above)
Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java]: > `rollupName` is null

done, changed.

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java (hunk 6 above)
Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java]: > modifyc -> modified?

done, changed.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableClauseVisitor.java (hunk 5 above)
Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableClauseVisitor.java:714]: better use `!table.isOlapTable()`

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/AddFieldClause.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    public String getColName() {
+        return colName;
+    }
+
+    public StructFieldDesc getFieldDesc() {
+        return fieldDesc;
+    }
+
+    public AddFieldClause(String colName, StructFieldDesc fieldDesc, String rollupName, Map<String, String> properties) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/AddFieldClause.java:36]: rollupName is useless, struct fields in rollup should be consistent with base table.

**Review by wyb — DISMISSED** [review]

**Review by gengjun-git — DISMISSED** [review]

**Review by packy92 — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        }
+
+        if (type.isArrayType()) {
+            return ((ArrayType) type).getItemType();
+        }
+
+        return null;
+    }
+
+    public void analyze(Column baseCol, boolean dropField) throws AnalysisException {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/StructFieldDesc.java:98]:
```suggestion
    public void analyze(Column baseCol, boolean dropField) throws SemanticException {
```

**Review by HangyuanLiu — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Comment by sevev** [comment]
@mergify backport branch-3.3

**Comment by sevev** [comment]
@mergify backport branch-3.3

**Comment by sevev** [comment]
@mergify backport branch-3.2

**Pull request merged**